[FLINK-38825] Add time-based batch triggering for AsyncBatchWaitOperator

featzhang opened the pull request:
### [FLINK-38825] Introduce AsyncBatchFunction and AsyncBatchWaitOperator (initial API & size-based batching)

#### Motivation

Many AI / ML inference and external service workloads benefit significantly from **batching** rather than per-record async invocation. Typical examples include:

* GPU-based model inference where batching improves throughput and utilization
* External services that expose batch APIs
* Databases or RPC systems with high per-request overhead

While Flink currently provides `AsyncFunction` and `AsyncWaitOperator` for record-level async I/O, there is no native abstraction for **batch-oriented async processing** in the DataStream API.

This PR introduces a **minimal and extensible foundation** for async batch processing in Flink.

---

#### What is included in this PR

This PR intentionally focuses on a **small, additive, and reviewer-friendly first step**:

1. **New public API**

   * Introduces `AsyncBatchFunction<IN, OUT>` (annotated with `@PublicEvolving`)
   * Allows users to implement async I/O over a batch of inputs

2. **New runtime operator**

   * Adds `AsyncBatchWaitOperator` with **unordered semantics**
   * Supports **size-based batching only**
   * Flushes remaining records on end-of-input
   * Keeps implementation intentionally simple

3. **Stream API entry point**

   * Adds `AsyncDataStream.unorderedWaitBatch(...)`
   * Mirrors existing async API style without modifying existing methods

4. **Unit tests**

   * Verifies batch-size triggering behavior
   * Verifies correct result emission
   * Verifies exception propagation and failure behavior

---

#### What is intentionally NOT included

To keep this PR minimal and avoid premature design constraints, the following are **explicitly out of scope** and will be addressed in follow-up PRs:

* Ordered batching semantics
* Time-based or event-time-based batching
* Retry / timeout handling at the async invocation level
* Metrics and backpressure tuning
* SQL / Table API / Python API integration
* Multiple in-flight batch concurrency control

---

#### Design notes

* The new API is **additive and backward-compatible**; no existing behavior is modified.
* The operator is designed to be easy to evolve, with clear extension points documented via TODOs.
* Implementation favors explicit logic over reuse or abstraction to simplify review and future changes.

---

#### Follow-up work

Planned follow-up PRs (tracked under FLINK-38825) include:

* Time-based batch triggering
* Ordered batch mode
* Event-time batching
* Retry / timeout strategies
* Metrics and observability
* Higher-level API and ecosystem integrations

---

#### Checklist

* [x] API annotated as `@PublicEvolving`
* [x] No changes to existing async APIs
* [x] Fully covered by unit tests
* [x] Backward-compatible and additive only

This PR establishes the minimal building block for async batch processing in Flink and enables iterative enhancement without locking in premature design decisions.

---

## PR Series for FLINK-38825

JIRA: [FLINK-38825](https://issues.apache.org/jira/browse/FLINK-38825) - Introduce an AI-friendly Async Batch Operator for high-latency inference workloads

This feature is implemented incrementally through the following PR series:

| # | PR | Title | Description | Module |
|---|---|---|---|---|
| 1 | [#27355](https://github.com/apache/flink/pull/27355) | Introduce AsyncBatchFunction and AsyncBatchWaitOperator | Core API and runtime operator with unordered semantics and size-based batch triggering | `flink-streaming-java` |
| **2** | [#27356](https://github.com/apache/flink/pull/27356) | Add time-based batch triggering | Timeout-based flush to trigger incomplete batches after a configurable duration | **`flink-streaming-java`** |
| 3 | [#27357](https://github.com/apache/flink/pull/27357) | Add ordered async batch support | Ordered output semantics via OrderedAsyncBatchWaitOperator with sequence numbers | `flink-streaming-java` |
| 4 | [#27358](https://github.com/apache/flink/pull/27358) | Add inference-oriented metrics | Batch size/latency histograms, async call duration, inflight batches, failure counters | `flink-streaming-java` |
| 5 | [#27359](https://github.com/apache/flink/pull/27359) | Implement retry and timeout strategies | Fixed-delay/exponential-backoff retry, fail-on-timeout/allow-partial timeout policies | `flink-streaming-java` |
| 6 | [#27360](https://github.com/apache/flink/pull/27360) | Add SQL/Table API integration | AsyncBatchLookupFunction, AsyncBatchLookupFunctionProvider, lookup join runner | `flink-table` |
| 7 | [#27361](https://github.com/apache/flink/pull/27361) | Add Python DataStream API integration | Python AsyncBatchFunction, async_invoke_batch() API, AsyncDataStream batch methods | `flink-python` |

> **Note**: Each PR builds incrementally on the previous ones. PRs should be reviewed and merged in order.

xishuaidelin commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/AsyncBatchWaitOperator.java:
```diff
@@ -0,0 +1,344 @@
+     * checks buffer state before flushing 2. Cancelling timers has overhead 3. Timer will be
+     * ignored if buffer is empty when it fires
+     */
+    private void clearTimerState() {
```
There might be an edge case here. Consider this scenario:
a. Element arrives at t=0, timer registered for t=100
b. Batch fills at t=10 → flushBuffer() clears state
c. New element arrives at t=20, timer registered for t=120
d. Old timer fires at t=100 → flushes the new partial batch unexpectedly

xishuaidelin commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/AsyncBatchWaitOperator.java:
```diff
@@ -0,0 +1,344 @@
+    private final transient MailboxExecutor mailboxExecutor;
+
+    /** Counter for in-flight async operations. */
+    private transient int inFlightCount;
```
flushBuffer() would increment inFlightCount but only endInput() waits for them. In a fast-producer with slow-consumer scenario, memory could grow fastly and unboundedly. How about set a max inflight count like AsyncWaitOperator.

xishuaidelin commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/async/AsyncBatchFunction.java:
```diff
@@ -0,0 +1,81 @@
+ * @param <OUT> The type of the returned elements.
+ */
+@PublicEvolving
+public interface AsyncBatchFunction<IN, OUT> extends Function, Serializable {
```
Serializable here is redundant.

xishuaidelin commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/AsyncBatchWaitOperator.java:
```diff
@@ -0,0 +1,344 @@
+            return;
+        }
+
+        // Clear timer state since we're flushing the batch
```
Nit: redundant comments

xishuaidelin commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/async/AsyncBatchWaitOperatorTest.java:
```diff
@@ -0,0 +1,517 @@
+        }
+    }
+
+    private static OneInputStreamOperatorTestHarness<Integer, Integer> createTestHarness(
```
Nit: this is equal to createTestHarnessWithTimeout(function, maxBatchSize, 0L).

xishuaidelin reviewed (commented):
Hi @featzhang. Thanks for your contribution. I have left some comments.

featzhang committed:
[FLINK-38825] Add AsyncBatchFunction and AsyncBatchWaitOperator with time-based batch triggering

This commit adds support for time-based batch triggering in async operations:
1. New AsyncBatchFunction interface for batch async I/O operations
2. AsyncBatchWaitOperator that batches incoming stream records
3. Support for both size-based and timeout-based batch triggers
4. Factory class for operator creation

Addresses reviewer comments:
- Fixed timer race condition by canceling pending timer in clearTimerState()
- Added capacity limit (default 100) for concurrent async batch operations
- Removed duplicate NO_TIMEOUT constant, using AsyncBatchWaitOperator.NO_TIMEOUT
- Removed redundant Serializable from AsyncBatchFunction interface
- Cleaned up redundant comments
- Fixed spotless formatting issues

featzhang committed:
[FLINK-38825] Prepare for spotless formatting check

featzhang committed:
[FLINK-38825] Fix private access to NO_TIMEOUT in AsyncBatchWaitOperatorFactory

Replace private constant AsyncBatchWaitOperator.NO_TIMEOUT with literal 0L
to fix compilation error in AsyncBatchWaitOperatorFactory constructor.

featzhang committed:
Refactor test code: make createTestHarness delegate to createTestHarnessWithTimeout

Address nit comment from @xishuaidelin by reusing the existing
createTestHarnessWithTimeout method with 0L timeout parameter,
reducing code duplication.

featzhang commented:
## Review Response Summary

Thank you @xishuaidelin for the thorough review! I've analyzed all 6 comments:

### Comments to Address:
1. **Timer edge case** - Old timer firing for new batch
2. **Unbounded inflight count** - Need max inflight limit
3. **Duplicate constant** - Consolidate definitions
4. **Redundant Serializable** - Remove from AsyncBatchFunction
5. **Redundant comments** - Clean up obvious comments
6. **Test helper simplification** - Reduce duplication

### Key Fixes:

**1. Timer Edge Case:**
```java
private Long currentTimerTimestamp = null;

private void flushBuffer() {
    // Cancel old timer to prevent race condition
    if (currentTimerTimestamp != null) {
        timerService.deleteProcessingTimeTimer(currentTimerTimestamp);
        currentTimerTimestamp = null;
    }
    // ... flush logic
}
```

**2. Max Inflight Limit:**
```java
private void flushBuffer() {
    while (inFlightCount >= maxInFlightBatches) {
        mailboxExecutor.yield();
    }
    inFlightCount++;
    // ... invoke async
}
```

**3. Remove Redundant Serializable:**
```java
// Function already extends Serializable
public interface AsyncBatchFunction<IN, OUT> extends Function {
    // No need for explicit Serializable
}
```

Full implementation guide available with test cases. Will update PR soon.

cc @featzhang
